draft-bbs-signatures.md

```diff
@@ -729,39 +735,60 @@ Procedure:
 
-The `hash_to_scalar` procedure hashes elements using an extendable-output function (xof). Applications not wishing to use an xof may use `hash_to_field` defined in Section 5.2 of [@!I-D.irtf-cfrg-hash-to-curve], combined with `expand_message_xmd` defined in Section 5.3.1 of the same document, in place of `hash_to_scalar`. In that case, every element outputted by `hash_to_field` that is equal to 0 MUST be rejected. If that occurs, one should calculate more field elements (using `hash_to_field`), until they get `n` non-zero elements (for example, if there is only one 0 in the output of `hash_to_field(msg, 2)` one must try to calculate `hash_to_field(msg, 3)` etc.).
+This operation makes use of expand\_message defined in [@!I-D.irtf-cfrg-hash-to-curve], in a similar way used by the hash\_to\_field operation of Section 5 from the same document (with the additional checks for getting a scalar that is 0). Note that, if an implementer wants to use hash\_to\_field here instead, they MUST use the multiplicative group of integers mod r (Fr), as the target group (F). However, the hash\_to\_curve ciphersuites used by this document, makes use of hash\_to\_field with the target group being the multiplicative group of integers mod p (Fp). For completeness, we define here the operation making use of the expand\_message function, that will be defined by the hash-to-curve ciphersuite used. If someone also has a hash\_to\_field implementation available, with the target group been Fr, they can use this instead (adding the check for a scalar been 0).
 
 ```
-result = hash_to_scalar(msg_octets, n)
+result = hash_to_scalar(msg_octets, count)
 
 Inputs:
 
 - msg_octets (REQUIRED), octet string. The message to be hashed.
-- n (REQUIRED), non-negative integer. The number of scalars to output.
+- count (REQUIRED), an integer greater or equal to 1. The number of
+                    scalars to output.
 
 Parameters:
 
-- r (REQUIRED), non-negative integer. The prime order of the G_1 and G_2 groups,
-     defined by the ciphersuite.
+- expand_message, as defined by the hash-to-curve suite uniquely
+                  identified by the hash_to_curve_id ciphersuite
+                  parameter.
+- dst, octet string. the separation tag defined by the
+       hash_to_curve_id ciphersuite parameter.
+- expand_length, non-negative integer. The number of bytes required to
+                 compute each scalar, defined by the ciphersuite.
 
 Outputs:
 
 - (scalar_1, ..., scalar_n), a list of non-zero scalars mod r.
 
 Procedure:
 
-1. h = xof(msg_octets)
+1. len_in_bytes = cound * expand_length
```

Comment:
I realized there's actually an issue with using expand_length here, as it conflicts with the recommended value for hash_to_field which is `ceil((ceil(log2(p)) + k) / 8)`. For BLS12-381 that is 48 bytes, but we have expand_length set to 64 bytes.

Reply:
Yes, 64 is the number of bytes when hashing to a G1 point (32 bytes are used for each component). If we want to allow the use of hash_to_field then I think we would need two different expand_length parameters, or we should just remove it from the ciphersuite and use definitions.

Reply:
Thank you very much for the feedback :) This is clearly confusing so moving to definitions is the best option IMO.

For `ProofGen` I removed the update from here and re-introduce it in PR #183. For `create_generators` and `hash_to_scalar` I made the update here.